### PR TITLE
Create Istio API Guidelines

### PR DESCRIPTION
**Pull request by wora**

### CODE REVIEW

**Comment by wora** [comment]
Fixed #316

File: STYLE.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+The Kubernetes CRD:
+
+```
+apiVersion: config.istio.io/v1alpha1
+kind: Authorization
+```
+
+The proto message definition:
+```
+package istio.config.v1alpha1;
```
Comment (kyessenov) [inline STYLE.md]: Should we use `io.istio.config` as prefix, to be precise?
We have folders here that correspond to feature areas. Should we map them to package, e.g:
```proto
package io.istio.api.routing.v1alpha1
message RouteRule { . . . }
```

**Comment by wora** [comment]
Across all major programming languages, only Java uses top-level domain in
the package name. Istio team can decide whether to include `io.` prefix.
Either way is fine, as long as we don't use both ways.

File: STYLE.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
+# Istio API Guidelines
+
+This page defines the design guidelines for Istio APIs. They apply to
+all proto files in this repository. Developers who develop their APIs
```
Comment (geeknoid) [inline STYLE.md]: Replace "in this repository" with "in the Istio project"

File: STYLE.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+using Istio may find these guidelines useful as well.
+
+Since Istio APIs are based on _proto3_ and _gRPC_, we will use
+Google's [API Design Guide](https://cloud.google.com/apis/design) as
+the baseline. Because Envoy APIs also uses the same baseline, the
+commonality across Envoy, Istio, proto3 and gRPC will greatly help
+developer experience in the long term.
+
+In addition, the following conventions should be followed for Istio
```
Comment (geeknoid) [inline STYLE.md]: In addition to Google's guide, the following...

File: STYLE.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+## Versioning
+
+When defining Kubernetes Custom Resource Definition (CRD) using
+proto3, the proto versioning should match the Kubernetes versioning,
+see the following example.
+
+The Kubernetes CRD:
+
+```
```
Comment (geeknoid) [inline STYLE.md]: Make that <tick><tick><tick>yaml, here and elsewhere.

File: STYLE.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
+# Istio API Guidelines
```
Comment (geeknoid) [inline STYLE.md]: Please rename the file to STYLE-GUIDE.md

Add a link to this file from the main README.md

File: STYLE.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+The Kubernetes CRD:
+
+```
+apiVersion: config.istio.io/v1alpha1
+kind: Authorization
+```
+
+The proto message definition:
+```
```
Comment (geeknoid) [inline STYLE.md]: Make that ```proto here and elsewhere

File: STYLE.md (hunk 1 above)
Comment (geeknoid) [reply STYLE.md]: We need to reorganize the repo I think such that the directory structure matches the package structure. Right now, it's a mess both in the repo and on the generated docs in istio.io

**Comment by wora** [comment]
I have addressed all comments. Thanks for reviewing.

File: STYLE-GUIDE.md [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+all proto files in the Istio project. Developers who develop their APIs
+using Istio may find these guidelines useful as well.
+
+Since Istio APIs are based on _proto3_ and _gRPC_, we will use
+Google's [API Design Guide](https://cloud.google.com/apis/design) as
+the baseline. Because Envoy APIs also uses the same baseline, the
+commonality across Envoy, Istio, proto3 and gRPC will greatly help
+developer experience in the long term.
+
+In addition to the Google's guide, the following conventions should be
```
Comment (geeknoid) [inline STYLE-GUIDE.md]: to the Google's -> to Google's

File: STYLE-GUIDE.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
+# Istio API Style Guide
+
+This page defines the design guidelines for Istio APIs. They apply to
+all proto files in the Istio project. Developers who develop their APIs
```
Comment (geeknoid) [inline STYLE-GUIDE.md]: who develop -> who create

**Comment by wora** [comment]
All fixed.

**Pull request merged**